prefetch.bs

```diff
@@ -225,19 +228,40 @@ A <dfn export>cross-origin prefetch IP anonymization policy</dfn> has an <dfn ex
   1. Return false.
 </div>
 
+A <dfn export>prefetch controller</dfn> is a [=struct=] which allows managing a prefetch operation. It has the following [=items=]:
+* <dfn for="prefetch controller">state</dfn> (default `"ongoing"`), which is `"ongoing"` or `"aborted"`
+
+  <div class="note">The state is `"ongoing"` even after prefetch is complete, as long as it was not aborted by the author.</div>
+* <dfn for="prefetch controller">fetch controller</dfn> (default null), which is a [=fetch controller=] or null
+* <dfn for="prefetch controller">prefetch record</dfn> (default null), which is a [=prefetch record=] or nulll
+* <dfn export for="prefetch controller">document</dfn>, which is a {{Document}}
+
+  <div class="note">This is the {{Document}} whose [=prefetch buffer=] contains this's [=prefetch controller/prefetch record=], if any.
```

Comment:
The existence of both Document's "prefetch buffer" and Document's "started prefetches" is pretty confusing right now. Similarly having all four structures (prefetch controller, started prefetch, prefetch record, prefetch candidate) is a bit tricky. Especially since this is all spread across two specs.

Here is my attempt at summarizing the situation and suggesting improvements...

- "prefetch candidate": result of parsing speculation rules. Sometimes stored persistently inside a "started prefetch", but it doesn't have identity, we're just keeping the URL and anonymization policy (not the referrer policy?) around for later comparison.
   - Suggestion: flatten URL and anonymization policy into "started prefetch", so that "prefetch candidate" is always the result of parsing and never has persistent existence.
   - Suggestion: rename to "prefetch speculation candidate" to reflect the strong connection to speculation rules? Unsure.
- "prefetch record": represents an ongoing prefetch, including eventually the response. Lives in "prefetch buffer".
- "started prefetch": represents an ongoing or aborted prefetch, including enough information to identify it, and to control (abort) it. Lives in "started prefetches"
  - Suggestion: merge with prefetch record?
- "prefetch controller": controls the prefetch. Only stored in "started prefetch". However it also has a reference to a "prefetch record".
  - Suggestion: remove the reference to "prefetch record" so that this is more of a leaf structure. Then, "abort" needs to take a prefetch record/started prefetch.
  - Possible suggestion: merge into "prefetch record" since it's not clear the independent existence gives us much...


Suggested end result, if possible:

- "prefetch speculation candidate" contains (URL, anonymization policy, referrer policy) and represents the result of parsing speculation rules.
- "Prefetch" takes a "prefetch speculation candidate" and turns it into a "prefetch record", which contains (URL, referrer policy, sandbox flags present, redirect chain, expiry time, isolated partition key, document, state?, fetch controller). "Prefetch" puts the resulting prefetch record into the document's "prefetch records"
- "Abort and discard" operates on "prefetch record". It aborts the fetch and also removes the record from the document's "prefetch records".

WDYT?

Reply:
Did more or less this refactoring. Excellent suggestion, thank you.